test/box-luatest/downgrade_test.lua

```diff
@@ -0,0 +1,799 @@
+            "Field constraint is found in space 'movement3d'," ..
+            " field 'azimuth'. It is supported starting from version 2.10.0.",
+        })
+    end)
+end
+
+-- Manual check:
+--
+--  Tarantool Enterprise 2.11.0-entrypoint-113-g803baaf
+--  type 'help' for interactive help
+--  tarantool> box.cfg{}
+--  tarantool> box.schema.func.create('in_box', {
+--      language = 'LUA',
+--      is_deterministic = true,
+--      body = [[function(row)
+--          return row.x > 0 and row.x < 1 and row.y > 0 and row.y < 1
+--      end]],
+--  })
+--  tarantool> box.schema.space.create('pos_in_box', {constraint = 'in_box'})
+--  tarantool> box.space.pos_in_box:format{{name = 'id', type = 'unsigned'},
+--                              {name = 'x', type = 'number'},
+--                              {name = 'y', type = 'number'}}
+--  tarantool> box.schema.downgrade('2.9.1')
+--  ---
+--  - error: 'builtin/box/upgrade.lua:1861: Tuple constraint is found in space
+--      ''pos_in_box''. It is supported starting from version 2.10.0.'
+--  ...
+g.test_downgrade_tuple_contraint = function(cg)
+    cg.server:exec(function()
+        local helper = require('test.box-luatest.downgrade_helper')
+
+        box.schema.func.create('in_box', {
+            language = 'LUA',
+            is_deterministic = true,
+            body = [[function(row)
+                return row.x > 0 and row.x < 1 and row.y > 0 and row.y < 1
+            end]],
+        })
+        local pos_in_box = box.schema.space.create('pos_in_box')
+        pos_in_box:format{{name = 'id', type = 'unsigned'},
+                          {name = 'x', type = 'number'},
+                          {name = 'y', type = 'number'}}
+        local flags = {constraint = {in_box = box.func.in_box.id}}
+        box.space._space:update(pos_in_box.id, {{'=', 'flags', flags}})
+
+        box.schema.func.create('in_circle', {
+            language = 'LUA',
+            is_deterministic = true,
+            body = [[function(row)
+                return row.x * row.x + row.y * row.y < 1
+            end]],
+        })
+        local pos_in_circle = box.schema.space.create('pos_in_circle')
+        pos_in_box:format{{name = 'id', type = 'unsigned'},
+                          {name = 'x', type = 'number'},
+                          {name = 'y', type = 'number'}}
+        local flags = {constraint = {in_circle = box.func.in_circle.id}}
+        box.space._space:update(pos_in_circle.id, {{'=', 'flags', flags}})
+
+        helper.check_issues('2.10.0', {
+            "Tuple constraint is found in space 'pos_in_box'. " ..
+            "It is supported starting from version 2.10.0.",
+            "Tuple constraint is found in space 'pos_in_circle'. " ..
+            "It is supported starting from version 2.10.0.",
+        })
+    end)
+end
+
+------------------------------
+-- Check downgrade from 2.10.5
+------------------------------
+
+-- No manual check required
```

Comment:
I don't understand, why are manual checks required above?

Reply:
Manual check is required to ensure we don't change system spaces in some unrelated way in test and then just handle this change in downgrade. For example `test_downgrade_tuple_contraint` can test for `conraint` instead of `constraint`. Hopefully API validates input and the described issue will be visible during manual test.

If we don't create anything artificial in test that no manual check is required.